sdk/macro/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+//! Convenience macro to declare a static public key and functions to interact with it
+//!
+//! bs58_string: bs58 string representation the program's id
+//!
+//! # Examples
+//!
+//! ```
+//! solana_sdk::declare_id!("My!!!11111111111111111111111111111111111111");
+//! ```
+
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use quote::{quote, ToTokens};
+use std::convert::TryFrom;
+use syn::{
+    parse::{Parse, ParseStream, Result},
+    parse_macro_input, LitByte, LitStr,
+};
+
+struct Id([u8; 32]);
+impl Parse for Id {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let id_literal: LitStr = input.parse()?;
+        if !input.is_empty() {
+            let stream: proc_macro2::TokenStream = input.parse()?;
+            return Err(syn::Error::new_spanned(stream, "unexpected token"));
+        }
+
+        let id_vec = bs58::decode(id_literal.value())
+            .into_vec()
+            .map_err(|_| syn::Error::new_spanned(&id_literal, "failed to decode base58 id"))?;
+        let id_array = <[u8; 32]>::try_from(<&[u8]>::clone(&&id_vec[..]))
+            .map_err(|_| syn::Error::new_spanned(&id_literal, "id is not 32 bytes long"))?;
```

Comment:
Minimal copy/paste needed to avoid cyclical dependency between solana-sdk and solana-sdk-macro